src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1336 @@
+
+#define _bsonVisit_applyOp(P, _const, _count)            \
+   do {                                                  \
+      if (!_bvContinue && !_bvHalt && !bsonParseError) { \
+         _bsonVisitOperation_##P;                        \
+      }                                                  \
+   } while (0);
+
+#define _bsonParse(Doc, ...)                                                   \
+   do {                                                                        \
+      /* Reset the context */                                                  \
+      struct _bsonVisitContext_t _bpCtx = {                                    \
+         .doc = &(Doc),                                                        \
+         .parent = &bsonVisitContext,                                          \
+      };                                                                       \
+      _bsonVisitContextThreadLocalPtr = &_bpCtx;                               \
+      /* Keep track of which elements have been visited based on their index*/ \
+      uint64_t _bpVisitBits_static[4] = {0};                                   \
+      BSON_MAYBE_UNUSED uint64_t *_bpVisitBits = _bpVisitBits_static;          \
+      BSON_MAYBE_UNUSED int _bpNumVisitBitInts =                               \
+         sizeof _bpVisitBits_static / sizeof (uint64_t);                       \
+      BSON_MAYBE_UNUSED bool _bpFoundElement = false;                          \
+      _bsonParse_applyOps (__VA_ARGS__);                                       \
+      /* We may have allocated for visit bits */                               \
+      if (_bpVisitBits != _bpVisitBits_static) {                               \
+         bson_free (_bpVisitBits);                                             \
+      }                                                                        \
+      /* Restore the dsl context */                                            \
+      _bsonVisitContextThreadLocalPtr = bsonVisitContext.parent;               \
+   } while (0)
+
+#define _bsonParse_applyOps(...) \
+   _bsonDSL_mapMacro (_bsonParse_applyOp, ~, __VA_ARGS__)
+
+/// Parse one entry referrenced by the context iterator
+#define _bsonParse_applyOp(P, _nil, Counter) \
+   do {                                      \
+      if (!_bvHalt && !bsonParseError) {     \
+         _bsonParseOperation_##P;            \
+      }                                      \
+   } while (0);
+
+#define _bsonParseMarkVisited(Index)                                   \
+   if (1) {                                                            \
+      const int nth_int = Index / 64;                                  \
+      const int nth_bit = Index % 64;                                  \
+      while (nth_int >= _bpNumVisitBitInts) {                          \
+         /* Say that five times, fast: */                              \
+         int new_num_visit_bit_ints = _bpNumVisitBitInts * 2;          \
```

Comment:
```suggestion
         size_t new_num_visit_bit_ints = _bpNumVisitBitInts * 2u;      \
```
Type consistency with `_bsonVisitContext_t::index`.